### PR TITLE
fix: Improve `instance` polling logic and introduce EventPoller class

### LINKED ISSUE #292: [Help]: Instance creation with custom disk sometimes fails
### Description

I'm creating instances (mainly in region `eu-central`) with custom disks and configs. Sometimes this fails with an error message. The full traceback (when running the playbook with `-vvvv`) is:

```
The full traceback is:
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/module_utils/linode_common.py", line 130, in __init__
    res = self.exec_module(**self.module.params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/modules/instance.py", line 856, in exec_module
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/modules/instance.py", line 818, in _handle_present
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/modules/instance.py", line 714, in _update_disks
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/modules/instance.py", line 550, in _create_disk_register
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/modules/instance.py", line 580, in _wait_for_disk_status
  File "/Users/[redacted]/.asdf/installs/python/3.11.2/lib/python3.11/site-packages/polling.py", line 106, in poll
    val = target(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/folders/m7/ykdfcv0x2xb1p75hztn5d9s40000gt/T/ansible_linode.cloud.instance_payload__eil645k/ansible_linode.cloud.instance_payload.zip/ansible_collections/linode/cloud/plugins/modules/instance.py", line 581, in <lambda>
  File "/Users/[redacted]/.asdf/installs/python/3.11.2/lib/python3.11/site-packages/linode_api4/objects/base.py", line 121, in __getattribute__
    self._api_get()
  File "/Users/[redacted]/.asdf/installs/python/3.11.2/lib/python3.11/site-packages/linode_api4/objects/base.py", line 205, in _api_get
    json = self._client.get(type(self).api_endpoint, model=self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/[redacted]/.asdf/installs/python/3.11.2/lib/python3.11/site-packages/linode_api4/linode_client.py", line 1547, in get
    return self._api_call(*args, method=self.session.get, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/[redacted]/.asdf/installs/python/3.11.2/lib/python3.11/site-packages/linode_api4/linode_client.py", line 1516, in _api_call
    raise ApiError(error_msg, status=response.status_code, json=j)
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "api_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "api_version": "v4",
            "authorized_keys": null,
            "backup_id": null,
            "boot_config_label": "boot-config",
            "booted": true,
            "configs": [
                {
                    "comments": null,
                    "devices": {
                        "sda": {
                            "disk_id": null,
                            "disk_label": "boot",
                            "volume_id": null
                        },
                        "sdb": {
                            "disk_id": null,
                            "disk_label": "minio-data",
                            "volume_id": null
                        },
                        "sdc": null,
                        "sdd": null,
                        "sde": null,
                        "sdf": null,
                        "sdg": null,
                        "sdh": null
                    },
                    "helpers": null,
                    "interfaces": null,
                    "kernel": null,
                    "label": "boot-config",
                    "memory_limit": null,
                    "root_device": "/dev/sda",
                    "run_level": null,
                    "virt_mode": null
                }
            ],
            "disks": [
                {
                    "authorized_keys": null,
                    "authorized_users": null,
                    "filesystem": null,
                    "image": "linode/ubuntu22.04",
                    "label": "boot",
                    "root_pass": "ans1ble-test!",
                    "size": 10000,
                    "stackscript_data": null,
                    "stackscript_id": null
                },
                {
                    "authorized_keys": null,
                    "authorized_users": null,
                    "filesystem": "raw",
                    "image": null,
                    "label": "minio-data",
                    "root_pass": null,
                    "size": 14000,
                    "stackscript_data": null,
                    "stackscript_id": null
                }
            ],
            "group": null,
            "image": null,
            "interfaces": null,
            "label": "foo",
            "private_ip": null,
            "region": "eu-central",
            "root_pass": null,
            "stackscript_data": null,
            "stackscript_id": null,
            "state": "present",
            "tags": null,
            "type": "g6-nanode-1",
            "ua_prefix": null,
            "wait": true,
            "wait_timeout": 240
        }
    },
    "msg": "Error from Linode API: [404] Not found"
}
```

When viewing the node in the Linode Cloud Manager web UI while the ansible command was running, I noticed is that the ansible command failed **while the disk was still resizing** from the default size (25GB) to my desired size (say 10GB). From my testing it seems that the chance of this failing may not the same for all data centers; I'm testing on `eu-central` mainly.

So my question is basically if should do it differently, or maybe this is a bug in this ansible collection because it should wait for this disk to be completely ready or something?

(On a side note: in the documentation at https://github.com/linode/ansible_linode/blob/main/docs/modules/instance.md, the example with disks and configs contains `state: present`, but I get an error when adding that, saying that `state` is not allowed there.)

### Ansible Configuration

```
ANSIBLE_PIPELINING(/Users/[redacted]/deployment/ansible/ansible.cfg) = True
CONFIG_FILE() = /Users/[redacted]/deployment/ansible/ansible.cfg
LOCALHOST_WARNING(/Users/[redacted]/deployment/ansible/ansible.cfg) = False
```

Ansible version: `core 2.14.2`
Linode.cloud collection: `0.14.0`

### Attempted Config

Playbook:

```
- name: Foo
  hosts: localhost
  vars:
    disk1:
      label: boot
      image: linode/ubuntu22.04
      size: 10000
      root_pass: ans1ble-test!
    disk2:
      label: data
      filesystem: raw
      size: 14000
    boot_config:
      label: boot-config
      root_device: /dev/sda
      devices:
        sda:
          disk_label: boot
        sdb:
          disk_label: data
  tasks:
    - name: Create linode
      linode.cloud.instance:
        label: 'foo'
        region: eu-central
        type: g6-nanode-1
        booted: true
        state: present
        boot_config_label: boot-config
        disks: "{{ [disk1, disk2] }}"
        configs: "{{ [boot_config] }}"
```

### PR DESCRIPTION
**Pull request by lgarber-akamai**
## 📝 Description

This change makes various improvements to the `instance` module polling logic. Additionally, it introduces an `EventPoller` class for simpler and more reliable Linode event polling.

Resolves #292 

## ✔️ How to Test

```
make TEST_ARGS="-v instance_config_disk" test
```

### CODE REVIEW

**Comment by lgarber-akamai** [comment]
/acctest sha=a7ea2235256069740c818785cb5dda59cb220e26 tests=instance_config_disk

**Comment by lgarber-akamai** [comment]
/acctest sha=77dd773b453700209f90fd76c52ec447aa70eab6 tests=instance_config_disk

**Review by jriddle-linode — APPROVED** [review]

File: plugins/module_utils/linode_event_poller.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+POLL_INTERVAL_SECONDS = 2
+
+
+class EventPoller:
+    """
+    EventPoller allows modules to dynamically poll for Linode events
+    """
+
+    def __init__(self, client: LinodeClient, entity_type: str, action: str, entity_id: int = 0):
```
Comment (zliang-akamai) [inline plugins/module_utils/linode_event_poller.py:21]: Wonder if it's better to make the default value of `_entity_id` be `None`.

File: plugins/module_utils/linode_event_poller.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+                nonlocal result_event
+                result_event = event
+
+                # Merge the new events into the cache
+                for event in result['data']:
+                    if event['id'] in self._previous_event_cache:
+                        continue
+
+                    self._previous_event_cache[event['id']] = event
```
Comment (zliang-akamai) [inline plugins/module_utils/linode_event_poller.py:76]: This loop seems a duplicated of the loop at line 64. Is it possible to merge and simplified the logic of them?

Comment (zliang-akamai) [reply plugins/module_utils/linode_event_poller.py:76]: Something in my brain looks like:
```python
        def poll_func():
            result = self._client.get('/account/events', filters=self._build_filter())
            event_found = False
            for event in result['data']:
                if event['id'] in self._previous_event_cache:
                    continue

                nonlocal result_event
                if not event_found:
                    result_event = event
                    event_found = True
                self._previous_event_cache[event['id']] = event

            return event_found
```

Comment (lgarber-akamai) [reply plugins/module_utils/linode_event_poller.py:76]: Hmm, come to think of it the second loop seems redundant anyway. Lemme fix that

**Review by zliang-akamai — APPROVED** [review]
Nice!

**Pull request merged**